bentoml/yatai/grpc_stream_utils.py

```diff
@@ -0,0 +1,138 @@
+        self.is_request = is_request
+        self.out_stream = FileStream()
+        self.tar = tarfile.TarFile(fileobj=self.out_stream, mode='w')
+
+    @staticmethod
+    def _stream_file_into_tar(tarinfo, tar, file_handler, buf_size):
+        out = tar.fileobj
+
+        # For iter(), the second argument, sentinel, is given, then object must
+        # be a callable object. The iterator created in this case will
+        # call object with no arguments for each call to its __next__()
+        # method; if the value returned is equal to sentinel, StopIteration
+        # will be raised, otherwise the value will be returned.
+        for b in iter(lambda: file_handler.read(buf_size), b''):
+            out.write(b)
+            yield
+        blocks, remainder = divmod(tarinfo.size, tarfile.BLOCKSIZE)
+        if remainder > 0:
+            out.write(tarfile.NUL * (tarfile.BLOCKSIZE - remainder))
+            blocks += 1
+        tar.offset += blocks * tarfile.BLOCKSIZE
+        yield
+
+    def create_request_or_response(self):
+        if self.is_request:
+            return UploadBentoRequest(
+                bento_name=self.bento_name,
+                bento_version=self.bento_version,
+                bento_bundle=self.out_stream.read_value(),
+            )
+        else:
+            return DownloadBentoResponse(bento_bundle=self.out_stream.read_value())
+
+    def close(self):
+        self.tar.close()
+        self.out_stream.close()
+
+    def __iter__(self):
+        if self.is_request:
+            yield UploadBentoRequest(
+                bento_name=self.bento_name, bento_version=self.bento_version
+            )
+        # Include length of separator.
+        prefix_len = len(self.directory_path) + len(os.path.sep)
+
+        # Add the directory path to tar
+        self.tar.add(name=self.directory_path, arcname='', recursive=False)
+
+        # Manually walk the directory and add to tarfile
+        for path, dirs, files in os.walk(self.directory_path):
+            arc_path = path[prefix_len:]
+
+            # Add files to tar. Add the tar info and then stream the file into tar
+            for f in files:
+                file_path = os.path.join(path, f)
+                with open(file_path, 'rb') as file:
+                    # Reading the file info and generate tarinfo from that.
+                    # No file content is added to the tar at this point
+                    tar_info = self.tar.gettarinfo(
+                        name=file_path, arcname=os.path.join(arc_path, f), fileobj=file,
+                    )
+                    self.tar.addfile(tar_info)
+                    # stream the file content into the tar
+                    for _ in self._stream_file_into_tar(
+                        tar_info, self.tar, file, self.file_chunk_size
+                    ):
+                        # Stream out as tarfile
+                        yield self.create_request_or_response()
```

Comment:
the code here is a bit hard to follow, can we make `create_request_or_response` a static helper method:

```
def create_request_or_response(bento_bundle_chunk_value):
  ...
```

And here we can have `yield create_request_or_response(self.tar.fileobj.read_value(), self.is_request)` ?